libs/execution_context/ebpf_program.c

```diff
@@ -182,61 +183,153 @@ _IRQL_requires_max_(PASSIVE_LEVEL) static ebpf_result_t _ebpf_program_compute_pr
     _In_ const uint32_t* actual_helper_ids,
     size_t count_of_actual_helper_ids,
     _In_ const ebpf_program_data_t* general_program_information_data,
-    _In_ const ebpf_program_data_t* type_specific_program_information_data,
+    _In_ const ebpf_program_data_t* extension_program_data,
     _In_ const cxplat_utf8_string_t* hash_algorithm,
     _Outptr_ uint8_t** hash,
     _Out_ size_t* hash_length);
 
+static bool
+_ebpf_program_match_provider_data_module_id(_In_ const PNPI_MODULEID npi_module_id, _In_ const GUID* expected_module_id)
+{
+    bool match = false;
+    // Verify the NPI module ID is a GUID.
+    if (npi_module_id->Type != MIT_GUID) {
+        EBPF_LOG_MESSAGE(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_PROGRAM,
+            "Program information provider module ID type mismatch.");
+        goto Done;
+    }
+
+    // Check if the provider module ID matches the expected module ID.
+    if (memcmp(&npi_module_id->Guid, expected_module_id, sizeof(GUID)) != 0) {
+        // This is expected as the attach callback will be called for each provider of NPI
+        // EBPF_PROGRAM_INFO_EXTENSION_IID.
+        goto Done;
+    }
+    match = true;
+Done:
+    return match;
+}
+
+enum ebpf_program_info_provider_type
+{
+    EBPF_PROGRAM_INFO_PROVIDER_TYPE_EXTENSION,
+    EBPF_PROGRAM_INFO_PROVIDER_TYPE_GENERAL_HELPER
+};
+
+static bool
+_ebpf_program_verify_provider_program_data(
+    _In_ enum ebpf_program_info_provider_type provider_type,
+    _In_ const PNPI_MODULEID npi_module_id,
+    _In_opt_ const ebpf_program_data_t* program_data)
+{
+    bool program_data_valid = false;
+    bool is_general_helper_program_data = (provider_type == EBPF_PROGRAM_INFO_PROVIDER_TYPE_GENERAL_HELPER);
+
+    // Verify that the program data is present.
+    if (program_data == NULL) {
+        EBPF_LOG_MESSAGE(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_PROGRAM,
+            "The program information provider data is not present.");
+        goto Done;
+    }
+
+    // Verify that that the program information version is supported.
+    if (program_data->header.version > EBPF_PROGRAM_DATA_VERSION_LATEST ||
```

Comment:
Suggest
```suggestion
    if (program_data->header.version != EBPF_PROGRAM_DATA_VERSION_LATEST ||
```

Since if VERSION_LATEST ever changes to 1 then this line will need to be updated, so it's best to cause a failure if that happens.